tests/phpunit/CRM/Civicase/Hook/SelectWhereClause/LimitCaseQueryToAccessibleCaseCategoriesTest.php

```diff
@@ -0,0 +1,140 @@
+      'case_type_category' => 2,
+    ]);
+    static::$client = ContactFabricator::fabricate();
+
+    CaseFabricator::fabricate(
+      [
+        'case_type_id' => static::$caseTypes[0]['id'],
+        'contact_id' => static::$client['id'],
+        'creator_id' => static::$client['id'],
+      ]
+    );
+    CaseFabricator::fabricate(
+      [
+        'case_type_id' => static::$caseTypes[1]['id'],
+        'contact_id' => static::$client['id'],
+        'creator_id' => static::$client['id'],
+      ]
+    );
+  }
+
+  /**
+   * Test that all records are returned if user has no permissions.
+   */
+  public function testAllCasesAreReturnedForNoPermissions() {
+    $contact = ContactFabricator::fabricate();
+    $this->registerCurrentLoggedInContactInSession($contact['id']);
+    $this->setPermissions(['access my cases and activities']);
+
+    $result = civicrm_api3('Case', 'get', [
+      'contact_id' => static::$client['id'],
+      'check_permissions' => 1,
+    ]);
+
+    $this->assertEquals($result['count'], 2);
+  }
+
+  /**
+   * Test that all records are returned if user has all permissions.
+   */
+  public function testAllCasesAreReturnedForAllPermissions() {
+    $contact = ContactFabricator::fabricate();
+    $this->registerCurrentLoggedInContactInSession($contact['id']);
+    $this->setPermissions(['basic case information', 'basic award information']);
+
+    $result = civicrm_api3('Case', 'get', [
+      'contact_id' => static::$client['id'],
+      'check_permissions' => 1,
+    ]);
+
+    $this->assertEquals($result['count'], 2);
+  }
+
+  /**
+   * Test that only accessible records are returned for limited permissions.
+   */
+  public function testOnlyAccessibleCasesAreReturnedForLimitedPermissions() {
+    $contact = ContactFabricator::fabricate();
+    $this->registerCurrentLoggedInContactInSession($contact['id']);
+    $this->setPermissions(['basic case information']);
+
+    $result = civicrm_api3('Case', 'get', [
+      'contact_id' => static::$client['id'],
+      'check_permissions' => 1,
+      'sequential' => 1,
+    ]);
+
+    $this->assertEquals($result['count'], 1);
+    $this->assertEquals($result['values'][0]['case_type_id'], static::$caseTypes[0]['id']);
+  }
+
+  /**
+   * Test that only accessible records are returned for limited permissions.
+   */
+  public function testOnlyAccessibleCasesAreReturnedForLimitedPermissionsWithCaseTypeCategoryParameter() {
+    $contact = ContactFabricator::fabricate();
+    $this->registerCurrentLoggedInContactInSession($contact['id']);
+    $this->setPermissions([
+      'basic case information',
+      'access CiviCRM',
+      'access my award and activities',
+    ]);
+
+    $result = civicrm_api3('Case', 'getcaselist', [
+      'contact_id' => static::$client['id'],
+      'case_type_id.case_type_category' => [
+        'IN' => ['award'],
+      ],
+      'check_permissions' => 1,
+      'sequential' => 1,
+    ]);
+
+    $this->assertEquals($result['count'], 0);
```

Comment:
Not sure about the test title. The title says `testOnlyAccessibleCasesAreReturned` but no results are returned. Maybe we can rename to  `testNoCasesAreReturnedWhenUserTriesToFetchCasesForCategoryItDoesNotHaveAccessTo` or some better title.

Also we need one more test where the user is able to fetch for the case category he has access to and we ensure that only such cases are returned and it does not leak into other case type categories when the case type category parameter is sent, i.e `testCasesAreReturnedOnlyForCasesUserHasAccessToWhenCaseCategoryParameterIsSent`